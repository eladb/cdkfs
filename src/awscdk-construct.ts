import * as path from 'path';
import { glob } from 'glob';
import * as semver from 'semver';
import { ConstructLibrary, ConstructLibraryOptions } from './construct-lib';
import { NodeLambdaConstructSourceCode } from './node-lambda-construct-source-code';


/**
 * Options for the construct-lib-aws project.
 */
export interface AwsCdkConstructLibraryOptions extends ConstructLibraryOptions {
  /**
   * Minimum target version this library is tested against.
   *
   * @default "1.95.2"
   * @featured
   */
  readonly cdkVersion: string;

  /**
   * Minimum target version of constructs being tested against. If not provided,
   * the default value depends on the configured `cdkVersion`:
   *
   * - For CDK 1.x, the default is "3.2.27"
   * - For CDK 2.x, the default is "10.0.5"
   * - Otherwise, the default is "*"
   *
   * @default - When the default behavior is used, the dependency on `constructs` will only
   * be added as a `peerDependency`. Otherwise, a `devDependency` will also be
   * added, set to the exact version configrued here.
   */
  readonly constructsVersion?: string;

  /**
   * Use pinned version instead of caret version for CDK.
   *
   * You can use this to prevent yarn to mix versions for your CDK dependencies and to prevent auto-updates.
   * If you use experimental features this will let you define the moment you include breaking changes.
   *
   * @default false
   */
  readonly cdkVersionPinning?: boolean;

  /**
   * Which AWS CDK modules (those that start with "@aws-cdk/") does this library
   * require when consumed?
   * @featured
   */
  readonly cdkDependencies?: string[];

  /**
   * If this is enabled (default), all modules declared in `cdkDependencies` will be also added as
   * normal `dependencies` (as well as `peerDependencies`).
   *
   * This is to ensure that downstream consumers actually have your CDK dependencies installed
   * when using npm < 7 or yarn, where peer dependencies are not automatically installed.
   * If this is disabled, `cdkDependencies` will be added to `devDependencies` to ensure
   * they are present during development.
   *
   * @default true
   */
  readonly cdkDependenciesAsDeps?: boolean;

  /**
   * Install the @aws-cdk/assert library?
   * @default true
   */
  readonly cdkAssert?: boolean;

  /**
   * AWS CDK modules required for testing.
   * @featured
   */
  readonly cdkTestDependencies?: string[];

  readonly bundleLambda?: boolean;

  readonly bundleLambdaOptions?: BundleLambdaOptions;
}

export interface BundleLambdaOptions {
  /**
   * Suffix to be used to scan for lambda handlers.
   */
  readonly suffix?: string;

  /**
   * Should generate a construct with the name of the lambda.
   */
  readonly generateConstruct?: boolean;
}

/**
 * AWS CDK construct library project
 *
 * A multi-language (jsii) construct library which vends constructs designed to
 * use within the AWS CDK with a friendly workflow and automatic publishing to
 * the construct catalog.
 *
 * ```ts
 * const project = new ConstructLibraryAws({
 *   name: 'cdk-watchful',
 *   description: 'Watching your CDK apps since 2019',
 *   jsiiVersion: Semver.caret('1.7.0'),
 *   authorName: 'Elad Ben-Israel',
 *   authorEmail: 'elad.benisrael@gmail.com',
 *   repository: 'https://github.com/eladb/cdk-watchful.git',
 *   keywords: [
 *     "cloudwatch",
 *     "monitoring"
 *   ],
 *
 *   catalog: {
 *     twitter: 'emeshbi'
 *   },
 *
 *   // creates PRs for projen upgrades
 *   projenUpgradeSecret: 'PROJEN_GITHUB_TOKEN',
 *
 *   cdkVersion: '1.54.0',
 *   cdkDependencies: [
 *     "@aws-cdk/aws-apigateway",
 *     "@aws-cdk/aws-cloudwatch",
 *     "@aws-cdk/aws-cloudwatch-actions",
 *     "@aws-cdk/aws-dynamodb",
 *     "@aws-cdk/aws-ecs",
 *     "@aws-cdk/aws-ecs-patterns",
 *     "@aws-cdk/aws-elasticloadbalancingv2",
 *     "@aws-cdk/aws-events",
 *     "@aws-cdk/aws-events-targets",
 *     "@aws-cdk/aws-lambda",
 *     "@aws-cdk/aws-rds",
 *     "@aws-cdk/aws-sns",
 *     "@aws-cdk/aws-sns-subscriptions",
 *     "@aws-cdk/aws-sqs",
 *     "@aws-cdk/core"
 *   ],
 *   devDependencies: {
 *     "aws-sdk": Semver.caret("2.708.0")
 *   },
 *
 *   // jsii publishing
 *
 *   java: {
 *     javaPackage: 'com.github.eladb.watchful',
 *     mavenGroupId: 'com.github.eladb',
 *     mavenArtifactId: 'cdk-watchful'
 *   },
 *   python: {
 *     distName: 'cdk-watchful',
 *     module: 'cdk_watchful'
 *   }
 * });
 *
 * project.synth();
 * ```
 *
 * @pjid awscdk-construct
 */
export class AwsCdkConstructLibrary extends ConstructLibrary {
  /**
   * The target CDK version for this library.
   */
  public readonly cdkVersion: string;

  /**
   * Whether CDK dependencies are added as normal dependencies (and peer dependencies).
   */
  public readonly cdkDependenciesAsDeps: boolean

  constructor(options: AwsCdkConstructLibraryOptions) {
    super({
      ...options,
      peerDependencyOptions: {
        pinnedDevDependency: false,
      },
    });

    this.cdkVersion = options.cdkVersionPinning ? options.cdkVersion : `^${options.cdkVersion}`;
    this.cdkDependenciesAsDeps = options.cdkDependenciesAsDeps ?? true;

    this.logger.info('hello!!!!');
    const cdkMajorVersion = semver.minVersion(this.cdkVersion)?.major ?? 1;
    if (options.constructsVersion) {
      this.addPeerDeps(`constructs@^${options.constructsVersion}`);
      this.addDevDeps(`constructs@${options.constructsVersion}`);
    } else if (cdkMajorVersion === 1) {
      // CDK 1.x is built on constructs 3.x
      this.addPeerDeps('constructs@^3.2.27');
    } else if (cdkMajorVersion == 2) {
      // CDK 2.x is built on constructs 10.x
      this.addPeerDeps('constructs@^10.0.5');
    } else {
      // Otherwise, let the user manage which version they use
      this.addPeerDeps('constructs');
    }

    if (options.cdkAssert ?? true) {
      this.addDevDeps(this.formatModuleSpec('@aws-cdk/assert'));
    }

    this.addCdkDependencies(...options.cdkDependencies ?? []);
    this.addCdkTestDependencies(...options.cdkTestDependencies ?? []);

    if (options.bundleLambda ?? true) {
      this.bundleLambdas(options.bundleLambdaOptions);
    }
  }

  /**
   * @deprecated use `cdkVersion`
   */
  public get version() {
    return this.cdkVersion;
  }

  /**
   * Adds CDK modules as runtime dependencies.
   *
   * Modules are currently by default added with a caret CDK version both as "dependencies"
   * and "peerDependencies". This is because currently npm would not
   * automatically install peer dependencies that are not declared as concerete
   * dependencies by the consumer, so this is a little npm "hack" so that
   * consumers will not need to depend on them directly if they don't interact
   * with them.
   * See `cdkDependenciesAsDeps` for changing the default behavior.
   *
   * @param deps names of cdk modules (e.g. `@aws-cdk/aws-lambda`).
   */
  public addCdkDependencies(...deps: string[]) {
    // this ugliness will go away in cdk v2.0
    this.addPeerDeps(...deps.map(m => this.formatModuleSpec(m)));

    if (this.cdkDependenciesAsDeps) {
      this.addDeps(...deps.map(m => this.formatModuleSpec(m)));
    } else {
      this.addDevDeps(...deps.map(m => this.formatModuleSpec(m)));
    }
  }

  /**
   * Adds CDK modules as test dependencies.
   *
   * @param deps names of cdk modules (e.g. `@aws-cdk/aws-lambda`).
   */
  public addCdkTestDependencies(...deps: string[]) {
    this.addDevDeps(...deps.map(m => this.formatModuleSpec(m)));
  }

  private formatModuleSpec(module: string): string {
    return `${module}@${this.cdkVersion}`;
  }

  /**
   * Enables CDK apps to consume construct library lambdas.
   *
   * @param bundleLambdaOptions Bundle lambda options
   * @private
   */
  private bundleLambdas(bundleLambdaOptions?: BundleLambdaOptions) {
    this.logger.debug('running bundleLambdas');
    const lambdaSuffix = (bundleLambdaOptions?.suffix ?? '.lambda');
    const generateConstruct = bundleLambdaOptions?.generateConstruct ?? true;
    const handlerFilePattern = `${this.srcdir}/**/*${lambdaSuffix}\\.@(ts|js)`;

    this.addDevDeps('aws-sdk'); // common lambda runtime dependency
    this.addDevDeps('esbuild');

    this.eslint?.addOverride({
      rules: {
        'import/no-extraneous-dependencies': 'off',
      },
      files: [handlerFilePattern],
    });

    // create bundle compile task for each lambda handler, generate convenience construct
    glob.sync(handlerFilePattern).forEach(file => {
      this.logger.debug(`found lambda handler ${file}`);
      const parsedPath = path.parse(file);
      const bundleDirName = `${parsedPath.name}.bundle`;
      const bundleDir = path.join(parsedPath.dir, `${parsedPath.name}.bundle`);
      const compiledHandlerPath = bundleDir.replace(this.srcdir, this.libdir);
      const constructFilePath = path.join(parsedPath.dir, parsedPath.name.replace(lambdaSuffix, '') + '.ts');

      if (generateConstruct) {
        new NodeLambdaConstructSourceCode(this, constructFilePath, bundleDirName);
      }
      const bundleLambdaTask = this.addTask(`bundleLambda:${file}`, {
        description: `Bundle lambda: ${file}`,
        exec: [
          'esbuild',
          '--bundle',
          file,
          '--target="node14"',
          '--platform="node"',
          `--outfile="${path.join(compiledHandlerPath, 'index.js')}"`,
          '--external:aws-sdk',
        ].join(' '),
      });
      this.compileTask.spawn(bundleLambdaTask);
    });
  }

}

/** @deprecated use `AwsCdkConstructLibraryOptions` */
export interface ConstructLibraryAwsOptions extends AwsCdkConstructLibraryOptions {
}

/** @deprecated use `AwsCdkConstructLibrary` */
export class ConstructLibraryAws extends AwsCdkConstructLibrary {
}
