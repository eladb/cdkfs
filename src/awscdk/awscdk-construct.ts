import * as semver from 'semver';
import { ConstructLibrary, ConstructLibraryOptions } from '../cdk';
import { AutoDiscover } from './auto-discover';
import { LambdaFunctionCommonOptions } from './lambda-function';

/**
 * Options for the construct-lib-aws project.
 */
export interface AwsCdkConstructLibraryOptions extends ConstructLibraryOptions {
  /**
   * Minimum target version this library is tested against.
   *
   * @default "2.0.0"
   * @featured
   */
  readonly cdkVersion: string;

  /**
   * Minimum target version of constructs being tested against. If not provided,
   * the default value depends on the configured `cdkVersion`:
   *
   * - For CDK 1.x, the default is "3.2.27"
   * - For CDK 2.x, the default is "10.0.5"
   * - Otherwise, the default is "*"
   *
   * @default - When the default behavior is used, the dependency on `constructs` will
   * be added as both a `peerDependency` and a `devDependency`.
   */
  readonly constructsVersion?: string;

  /**
   * Use pinned version instead of caret version for CDK.
   *
   * You can use this to prevent yarn to mix versions for your CDK dependencies and to prevent auto-updates.
   * If you use experimental features this will let you define the moment you include breaking changes.
   *
   * @default false
   */
  readonly cdkVersionPinning?: boolean;

  /**
   * Which AWS CDK modules (those that start with "@aws-cdk/") does this library
   * require when consumed?
   *
   * @deprecated - For CDK 2.x use 'peerDeps' instead
   */
  readonly cdkDependencies?: string[];

  /**
   * If this is enabled (default), all modules declared in `cdkDependencies` will be also added as
   * normal `dependencies` (as well as `peerDependencies`).
   *
   * This is to ensure that downstream consumers actually have your CDK dependencies installed
   * when using npm < 7 or yarn, where peer dependencies are not automatically installed.
   * If this is disabled, `cdkDependencies` will be added to `devDependencies` to ensure
   * they are present during development.
   *
   * @default true
   * @deprecated - Not used for CDK 2.x
   */
  readonly cdkDependenciesAsDeps?: boolean;

  /**
   * Install the @aws-cdk/assert library?
   *
   * @deprecated - use cdkAssertions instead
   * @default false
   */
  readonly cdkAssert?: boolean;

  /**
   * Install the @aws-cdk/assertions library?
   *
   * Only needed for CDK 1.x. If using CDK 2.x then
   * assertions is already included in 'aws-cdk-lib'
   *
   * @default true
   */
  readonly cdkAssertions?: boolean;

  /**
   * AWS CDK modules required for testing.
   *
   *
   *
   * @featured
   * @deprecated - For CDK 2.x use 'devDeps' instead
   */
  readonly cdkTestDependencies?: string[];

  /**
   * Automatically adds an `aws_lambda.Function` for each `.lambda.ts` handler
   * in your source tree. If this is disabled, you either need to explicitly
   * call `aws_lambda.Function.autoDiscover()` or define a `new
   * aws_lambda.Function()` for each handler.
   *
   * @default true
   */
  readonly lambdaAutoDiscover?: boolean;

  /**
    * Common options for all AWS Lambda functions.
    *
    * @default - default options
    */
  readonly lambdaOptions?: LambdaFunctionCommonOptions;
}

/**
 * AWS CDK construct library project
 *
 * A multi-language (jsii) construct library which vends constructs designed to
 * use within the AWS CDK with a friendly workflow and automatic publishing to
 * the construct catalog.
 *
 * @pjid awscdk-construct
 */
export class AwsCdkConstructLibrary extends ConstructLibrary {
  /**
   * The target CDK version for this library.
   */
  public readonly cdkVersion: string;

  /**
   * Whether CDK dependencies are added as normal dependencies (and peer dependencies).
   */
  public readonly cdkDependenciesAsDeps: boolean;

  constructor(options: AwsCdkConstructLibraryOptions) {
    const cdkVersion = options.cdkVersionPinning ? options.cdkVersion : `^${options.cdkVersion}`;
    const cdkMajorVersion = semver.minVersion(cdkVersion)?.major ?? 2;
    super({
      peerDependencyOptions: cdkMajorVersion === 1 ? {
        pinnedDevDependency: false,
      } : undefined,
      ...options,
    });

    this.cdkVersion = cdkVersion;

    this.cdkDependenciesAsDeps = options.cdkDependenciesAsDeps ?? true;

    switch (cdkMajorVersion) {
      case 1:
        if (options.cdkAssert) {
          this.addDevDeps(this.formatModuleSpec('@aws-cdk/assert'));
        }
        if (options.cdkAssertions ?? true) {
          this.addDevDeps(this.formatModuleSpec('@aws-cdk/assertions'));
        }
        // CDK 1.x is built on constructs 3.x
        if (options.constructsVersion) {
          const constructsMajorVersion = semver.minVersion(options.constructsVersion)?.major;
          if (constructsMajorVersion !== undefined && constructsMajorVersion >= 10) {
            throw new Error('CDK 1.x requires constructs version < 10');
          }
          this.addPeerDeps(`constructs@^${options.constructsVersion}`);
        } else {
          this.addPeerDeps('constructs@^3.2.27');
        }
        this.addCdkDependencies(...options.cdkDependencies ?? []);
        this.addCdkTestDependencies(...options.cdkTestDependencies ?? []);
        break;
      case 2:
        if (options.cdkAssert !== undefined) {
          throw new Error('cdkAssert is not used for CDK 2.x. Use the assertions library that is provided in aws-cdk-lib');
        }
        if (options.cdkDependencies !== undefined) {
          throw new Error('cdkDependencies is not used for CDK 2.x. Use "peerDeps" instead');
        }
        if (options.cdkDependenciesAsDeps !== undefined) {
          throw new Error('cdkDependenciesAsDeps is not used for CDK 2.x');
        }
        if (options.cdkTestDependencies !== undefined) {
          throw new Error('cdkTestDependencies is not used for CDK 2.x. Use "devDeps" instead');
        }
        // CDK 2.x is built on constructs 10.x
        if (options.constructsVersion) {
          const constructsMajorVersion = semver.minVersion(options.constructsVersion)?.major;
          if (constructsMajorVersion !== undefined && constructsMajorVersion < 10) {
            throw new Error('CDK 2.x requires constructs version >= 10');
          }
          this.addPeerDeps(`constructs@^${options.constructsVersion}`);
        } else {
          this.addPeerDeps('constructs@^10.0.5');
        }
        this.addPeerDeps(`aws-cdk-lib@${this.cdkVersion}`);

        break;
      default:
        // Otherwise, let the user manage which version they use
        this.addPeerDeps('constructs');
        break;
    }


    const lambdaAutoDiscover = options.lambdaAutoDiscover ?? true;
    if (lambdaAutoDiscover) {
      new AutoDiscover(this, {
        srcdir: this.srcdir,
        testdir: this.testdir,
        lambdaOptions: options.lambdaOptions,
        tsconfigPath: this.tsconfigDev.fileName,
      });
    }
  }

  /**
   * @deprecated use `cdkVersion`
   */
  public get version() {
    return this.cdkVersion;
  }

  /**
   * Adds CDK modules as runtime dependencies.
   *
   * Modules are currently by default added with a caret CDK version both as "dependencies"
   * and "peerDependencies". This is because currently npm would not
   * automatically install peer dependencies that are not declared as concerete
   * dependencies by the consumer, so this is a little npm "hack" so that
   * consumers will not need to depend on them directly if they don't interact
   * with them.
   * See `cdkDependenciesAsDeps` for changing the default behavior.
   *
   * @param deps names of cdk modules (e.g. `@aws-cdk/aws-lambda`).
   */
  public addCdkDependencies(...deps: string[]) {
    // this ugliness will go away in cdk v2.0
    this.addPeerDeps(...deps.map(m => this.formatModuleSpec(m)));

    if (this.cdkDependenciesAsDeps) {
      this.addDeps(...deps.map(m => this.formatModuleSpec(m)));
    } else {
      this.addDevDeps(...deps.map(m => this.formatModuleSpec(m)));
    }
  }

  /**
   * Adds CDK modules as test dependencies.
   *
   * @param deps names of cdk modules (e.g. `@aws-cdk/aws-lambda`).
   */
  public addCdkTestDependencies(...deps: string[]) {
    this.addDevDeps(...deps.map(m => this.formatModuleSpec(m)));
  }

  private formatModuleSpec(module: string): string {
    return `${module}@${this.cdkVersion}`;
  }
}

/** @deprecated use `AwsCdkConstructLibraryOptions` */
export interface ConstructLibraryAwsOptions extends AwsCdkConstructLibraryOptions { }

/** @deprecated use `AwsCdkConstructLibrary` */
export class ConstructLibraryAws extends AwsCdkConstructLibrary { }
