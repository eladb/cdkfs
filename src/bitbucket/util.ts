

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}