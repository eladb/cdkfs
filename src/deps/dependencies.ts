import * as path from 'path';
import { PROJEN_DIR } from '../common';
import { Component } from '../component';
import { JsonFile } from '../json';
import { Project } from '../project';
import { Dependency, DependencyCoordinates, DependencyType, DepsManifest } from './model';

/**
 * The `Dependencies` component is responsible to track the list of dependencies
 * a project has, and then used by project types as the model for rendering
 * project-specific dependency manifests such as the dependencies section
 * `package.json` files.
 *
 * To add a dependency you can use a project-type specific API such as
 * `nodeProject.addDeps()` or use the generic API of `project.deps`:
 */
export class Dependencies extends Component {
  /**
   * The project-relative path of the deps manifest file.
   */
  public static readonly MANIFEST_FILE = path.posix.join(PROJEN_DIR, 'deps.json');

  /**
   * Returns the coordinates of a dependency spec.
   *
   * Given `foo@^3.4.0` returns `{ name: "foo", version: "^3.4.0" }`.
   * Given `bar@npm:@bar/legacy` returns `{ name: "bar", version: "npm:@bar/legacy" }`.
   */
  public static parseDependency(spec: string): DependencyCoordinates {
    const scope = spec.startsWith('@');
    if (scope) {
      spec = spec.substr(1);
    }

    const [module, ...version] = spec.split('@');
    const name = scope ? `@${module}` : module;
    if (version.length == 0) {
      return { name };
    } else {
      return { name, version: version?.join('@') };
    }
  }

  private readonly _deps = new Array<Dependency>();

  /**
   * Adds a dependencies component to the project.
   * @param project The parent project
   */
  constructor(project: Project) {
    super(project);

    // this is not really required at the moment, but actually quite useful as a
    // checked-in source of truth for dependencies and will potentially be
    // valuable in the future for CLI tools.
    new JsonFile(project, Dependencies.MANIFEST_FILE, {
      omitEmpty: true,
      obj: () => this.toJson(),
    });
  }

  /**
   * A copy of all dependencies recorded for this project.
   *
   * The list is sorted by type->name->version
   */
  public get all(): Dependency[] {
    return [...this._deps].sort(compareDeps).map(normalizeDep);
  }

  /**
   * Returns a dependency by name.
   *
   * Fails if there is no dependency defined by that name or if `type` is not
   * provided and there is more then one dependency type for this dependency.
   *
   * @param name The name of the dependency
   * @param type The dependency type. If this dependency is defined only for a
   * single type, this argument can be omitted.
   *
   * @returns a copy (cannot be modified)
   */
  public getDependency(name: string, type?: DependencyType): Dependency {
    const idx = this.tryGetDependencyIndex(name, type);
    if (idx === -1) {
      const msg = type
        ? `there is no ${type} dependency defined on "${name}"`
        : `there is no dependency defined on "${name}"`;

      throw new Error(msg);
    }

    return {
      ...normalizeDep(this._deps[idx]),
    };
  }

  /**
   * Adds a dependency to this project.
   * @param spec The dependency spec in the format `MODULE[@VERSION]` where
   * `MODULE` is the package-manager-specific module name and `VERSION` is an
   * optional semantic version requirement (e.g. `^3.4.0`).
   * @param type The type of the dependency.
   */
  public addDependency(spec: string, type: DependencyType, metadata: { [key: string]: any } = { }): Dependency {
    this.project.logger.debug(`${type}-dep ${spec}`);

    const dep: Dependency = {
      ...Dependencies.parseDependency(spec),
      type,
      metadata,
    };

    const existingDepIndex = this.tryGetDependencyIndex(dep.name, type);

    if (existingDepIndex !== -1) {
      const existingDep = this._deps[existingDepIndex];
      if (dep.version === existingDep.version || existingDep.version === undefined) {
        this.project.logger.debug(`updating existing ${dep.type}-dep ${dep.name} with more specific version/metadata`);
        this._deps[existingDepIndex] = dep;
      } else {
        throw new Error(`"${dep.name}" is already specified with different version: ${existingDep.version}`);
      }
    } else {
      this._deps.push(dep);
    }

    return dep;
  }

  /**
   * Removes a dependency.
   * @param name The name of the module to remove (without the version)
   * @param type The dependency type. This is only required if there the
   * dependency is defined for multiple types.
   */
  public removeDependency(name: string, type?: DependencyType) {
    const removeIndex = this.tryGetDependencyIndex(name, type);
    if (removeIndex === -1) {
      return;
    }

    this._deps.splice(removeIndex, 1);
  }

  private tryGetDependencyIndex(name: string, type?: DependencyType): number {
    const deps = this._deps.filter(d => d.name === name);
    if (deps.length === 0) {
      return -1; // not found
    }

    if (!type) {
      if (deps.length > 1) {
        throw new Error(`"${name}" is defined for multiple dependency types: ${deps.map(d => d.type).join(',')}. Please specify dependency type`);
      }

      type = deps[0].type;
    }

    return this._deps.findIndex(dep => dep.name === name && dep.type === type);
  }

  private toJson(): DepsManifest | undefined {
    if (this._deps.length === 0) { return undefined; }
    return {
      dependencies: this._deps.sort(compareDeps).map(normalizeDep),
    };
  }
}

function normalizeDep(d: Dependency) {
  const obj: any = { };
  for (const [k, v] of Object.entries(d)) {
    if (v == undefined) {continue;}
    if (typeof(v) === 'object' && Object.keys(v).length === 0) {continue;}
    if (Array.isArray(v) && v.length === 0) {continue;}
    obj[k] = v;
  }

  return obj;
}

function compareDeps(d1: Dependency, d2: Dependency) {
  return specOf(d1).localeCompare(specOf(d2));

  function specOf(dep: Dependency) {
    let spec = dep.type + ':' + dep.name;
    if (dep.version) {
      spec += '@' + dep.version;
    }
    return spec;
  }
}

