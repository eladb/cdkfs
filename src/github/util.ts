export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
