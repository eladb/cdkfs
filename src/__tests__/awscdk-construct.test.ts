import each from 'jest-each';
import { AwsCdkConstructLibrary, AwsCdkConstructLibraryOptions } from '../awscdk-construct';
import { DependencyType } from '../deps';
import { LogLevel } from '../logger';
import { NpmAccess } from '../node-package';
import { mkdtemp, synthSnapshot } from './util';

describe('constructs dependency selection', () => {
  test('user-selected', () => {
    // GIVEN
    const project = new TestProject({ cdkVersion: '1.100.0', constructsVersion: '42.1337.0-ultimate.∞' });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot['package.json']?.peerDependencies?.constructs).toBe('^42.1337.0-ultimate.∞');
    expect(snapshot['package.json']?.devDependencies?.constructs).toBe('42.1337.0-ultimate.∞');
    expect(snapshot['package.json']?.dependencies?.constructs).toBeUndefined();
  });


  test('for cdk 1.x', () => {
    // GIVEN
    const project = new TestProject({ cdkVersion: '1.100.0' });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot['package.json']?.peerDependencies?.constructs).toMatch(/^\^3\./);
    expect(snapshot['package.json']?.devDependencies?.constructs).toBeUndefined();
    expect(snapshot['package.json']?.dependencies?.constructs).toBeUndefined();
  });

  test('for cdk 2.x', () => {
    // GIVEN
    const project = new TestProject({ cdkVersion: '2.0.0-alpha.5' });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot['package.json']?.peerDependencies?.constructs).toMatch(/^\^10./);
    expect(snapshot['package.json']?.devDependencies?.constructs).toBeUndefined();
    expect(snapshot['package.json']?.dependencies?.constructs).toBeUndefined();
  });

  test('for cdk 3.x (does not exist yet)', () => {
    // GIVEN
    const project = new TestProject({ cdkVersion: '3.1337.42' });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot['package.json']?.peerDependencies?.constructs).toBe('*');
    expect(snapshot['package.json']?.devDependencies?.constructs).toBeUndefined();
    expect(snapshot['package.json']?.dependencies?.constructs).toBeUndefined();
  });
});

describe('lambda bundle', () => {

  const cdkVersion = '1.100.0';

  each(['ts', 'js']).test('given %s handler then construct and bundle task are created', (handlerExt: string) => {
    // GIVEN
    const project = new TestProject({
      cdkVersion,
    });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN construct created
    expect(snapshot[`src/__tests__/integration/lambda/mylambda-${handlerExt}.ts`]).toMatchSnapshot();
    expect(snapshot[`src/__tests__/integration/lambda/mylambda-${handlerExt}.ts`].length).toBeGreaterThan(1);

    // THEN bundle task created
    expect(project.compileTask.steps).toContainEqual(expect.objectContaining({
      spawn: `bundleLambda:src/__tests__/integration/lambda/mylambda-${handlerExt}.lambda.${handlerExt}`,
    }));
  });

  test('given custom suffix handler then construct and bundle task are created', () => {
    // GIVEN
    const project = new TestProject({
      cdkVersion,
      bundleLambdaOptions: {
        suffix: '.customlambda',
      },
    });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN construct created
    expect(snapshot['src/__tests__/integration/lambda/custom-suffix.ts'].length).toBeGreaterThan(1);

    // THEN bundle task created
    expect(project.compileTask.steps).toContainEqual(expect.objectContaining({
      spawn: 'bundleLambda:src/__tests__/integration/lambda/custom-suffix.customlambda.ts',
    }));
  });

  test('given then aws-sdk and esbuild devDeps added', () => {
    // GIVEN
    const project = new TestProject({
      cdkVersion,
    });

    // THEN
    expect(project.deps.getDependency('aws-sdk', DependencyType.BUILD)).toBeDefined();
    expect(project.deps.getDependency('esbuild', DependencyType.BUILD)).toBeDefined();
  });

  test('given then eslint rule override added', () => {
    // GIVEN
    const project = new TestProject({
      cdkVersion,
    });

    // THEN
    expect(project.eslint?.overrides).toContainEqual({
      rules: {
        'import/no-extraneous-dependencies': 'off',
      },
      files: [
        'src/**/*.lambda\\.@(ts|js)',
      ],
    });
  });

  test('do nothing when turned off', () => {
    // GIVEN
    const project = new TestProject({
      cdkVersion,
      bundleLambda: false,
    });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot['src/__tests__/integration/lambda/mylambda-ts.ts']).toBeUndefined();
    expect(project.compileTask.steps).toContainEqual(expect.not.objectContaining({
      spawn: 'bundleLambda:src/__tests__/integration/lambda/mylambda-ts.lambda.ts',
    }));
  });

  test('do not generate construct when option turned off', () => {
    // GIVEN
    const project = new TestProject({
      cdkVersion,
      bundleLambdaOptions: {
        generateConstruct: false,
      },
    });

    // WHEN
    const snapshot = synthSnapshot(project);

    // THEN
    expect(snapshot['src/__tests__/integration/lambda/mylambda-ts.ts']).toBeUndefined();
  });

});

const defaultOptions = {
  author: 'Nobody',
  authorAddress: 'nobody@nowhere.com',
  clobber: false,
  defaultReleaseBranch: 'main',
  jest: false,
  name: 'test-project',
  npmAccess: NpmAccess.PUBLIC,
  repositoryUrl: 'https://github.com/projen/projen.git',
} as const;

class TestProject extends AwsCdkConstructLibrary {
  constructor(options: Omit<AwsCdkConstructLibraryOptions, keyof typeof defaultOptions>) {
    super({
      outdir: mkdtemp(),
      logging: {
        level: LogLevel.OFF,
      },
      ...defaultOptions,
      ...options,
    });
  }
}
