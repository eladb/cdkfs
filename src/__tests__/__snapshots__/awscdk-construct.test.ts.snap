// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lambda bundle given js handler then construct and bundle task are created 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from 'constructs';

export interface MylambdaJsProps extends lambda.FunctionOptions {
}

export class MylambdaJs extends lambda.Function {
  constructor(scope: Construct, id: string, props: MylambdaJsProps = {}) {
    super(scope, id, {
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '/mylambda-js.lambda.bundle')),
      ...props,
    });
  }
}"
`;

exports[`lambda bundle given ts handler then construct and bundle task are created 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from 'constructs';

export interface MylambdaTsProps extends lambda.FunctionOptions {
}

export class MylambdaTs extends lambda.Function {
  constructor(scope: Construct, id: string, props: MylambdaTsProps = {}) {
    super(scope, id, {
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '/mylambda-ts.lambda.bundle')),
      ...props,
    });
  }
}"
`;
